### PR TITLE
Add Hivemind support for Rust

### PR DESCRIPTION
**Pull request by ViliamVadocz**
Explanation for how it works is on my wiki: https://github.com/ViliamVadocz/Hivemind/wiki

I am making a PR to RLBot/RLBot as well that makes it work. If you want to try it out, feel free to clone my Hivemind repo which has the modified framework as well as some examples.

Please be critical. I would not consider myself good at Rust, so I might have made mistakes.

Still to add:

- Documentation comments for functions

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -130,6 +130,7 @@
 pub use crate::{
     framework::{parse_framework_args, run_bot, Bot, FrameworkArgs},
     game::*,
+    hive::{parse_hive_framework_args, run_hive, HiveFrameworkArgs, Hivemind},
```
Comment (saintdev) [inline src/lib.rs]: I think only exporting the hive module would be a better idea. I'm open to leaving these exposed in the root if anyone feels strongly about it. @Quetzal2 ?

File: src/hive.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        let inputs = hive.tick(&packet);
+
+        match inputs.len().cmp(&drone_indices.len()) {
+            Ordering::Equal => {
+                for i in 0..drone_indices.len() {
+                    // first input corresponds to first index in drone_indices, etc.
+                    let drone = drone_indices[i];
+                    let input = &inputs[i];
+                    rlbot.update_player_input(drone, input)?;
+                }
```
[the remark this answers is not in the document]
Comment (ViliamVadocz) [reply src/hive.rs]: > Make this a method on RLBot. Maybe have it take a slice of tuples AsRef<[(index, &ControllerState)]>?

What do you mean?

File: src/hive.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+) -> Result<Option<HiveFrameworkArgs>, ()> {
+    if args.next().as_ref().map(|s| &s[..]) != Some("--rlbot-version") {
+        return Ok(None); // not launched by the framework
+    }
+    let rlbot_version = args.next().ok_or(())?;
+
+    if args.next().as_ref().map(|s| &s[..]) != Some("--rlbot-dll-directory") {
+        return Err(());
+    }
+    let rlbot_dll_directory = PathBuf::from(args.next().ok_or(())?);
```
[the remark this answers is not in the document]
Comment (ViliamVadocz) [reply src/hive.rs]: This is proving to be more difficult than I expected

File: src/hive.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    pub drone_indices: Vec<i32>,
+
+    _non_exhaustive: (),
+}
+
+impl From<HiveFrameworkArgs> for InitOptions {
+    fn from(args: HiveFrameworkArgs) -> Self {
+        Self::new().rlbot_dll_directory(args.rlbot_dll_directory)
+    }
+}
```
[the remark this answers is not in the document]
Comment (ViliamVadocz) [reply src/hive.rs]: Does not seem like that is the case?

File: src/lib.rs (hunk 1 above)
Comment (Quetzal2) [reply src/lib.rs]: I liked it like that, having the same use as `framework`, thus the same way to use it. But if we say that in general, people will be more likely to create bots than hiveminds, and that hivemind is more a feature, I would see how it'd make sense to not expose it like that.
I think it's more a design choice, and both are ok to me. Anyways, if we go with the compact version, the user would just have to `use rlbot::hive::{…}` to have them exposed like again.

File: src/hive.rs [hunk 5]
```diff
@@ -20,49 +21,57 @@ pub trait Hivemind {
...
+                println!(
+                    "hivemind tried sending inputs to an index that is not in it's drone indices"
+                );
             }
-            Ordering::Less => return Err(Box::<dyn Error>::from("too few inputs for drones")),
-            Ordering::Greater => return Err(Box::<dyn Error>::from("too many inputs for drones")),
         }
+
+        // Sending valid inputs to drones.
+        rlbot.update_multiple_inputs(valid_inputs.into_iter())?;
```
[the remark this answers is not in the document]
Comment (ViliamVadocz) [reply src/hive.rs]: Done

File: src/framework.rs [hunk 6]
```diff
@@ -117,6 +113,23 @@ fn parse_framework_command_line(
     }))
 }
 
+/// Parse the version and dll directory arguments passed by the framework.
+pub fn parse_version_and_directory(
```
[the remark this answers is not in the document]
Comment (ViliamVadocz) [reply src/framework.rs:117]: Done

File: src/hive.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+/// Your hivemind bot should implement this trait.
+/// Use this like you would use the Bot trait.
+///
+/// The drone_indices are indices of the bots you control.
+pub trait Hivemind {
+    /// This method is called after your hivemind has been passed into
+    /// `run_hive()`. It gives you the a vector of bot indices which
+    /// correspond to those which you control.
+    fn set_drone_indices(&mut self, indices: Vec<usize>);
```
[the remark this answers is not in the document]
Comment (ViliamVadocz) [reply src/hive.rs:19]: I don't know enough about how this works to implement it. When I tried your change it broke the existing example bot I have in my Hivemind repo and I couldn't get it to work.

**Pull request merged**